LanguageFeatures/Patterns/type_inference_A29_t06.dart

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion
+/// To type check a pattern p being matched against a value of type M:
+/// ...
+/// Map:
+/// i. Calculate the value's entry key type K and value type V, and key context
+///     C:
+///   a. If p has type arguments <K, V> for some K and V then use those, and C
+///     is K.
+///   b. Else if M implements Map<K, V> for some K and V then use those, and C
+///     is K.
+///   c. Else if M is dynamic then K and V are dynamic and C is _.
+///   d. Else K and V are Object? and C is _.
+/// ii. Type-check each key expression using C as the context type.
+/// iii. Type-check each value subpattern using V as the matched value type.
+/// vi. The required type of p is Map<K, V>.
+///
+/// @description Check that if p has a type arguments <K1, V1> and M implements
+/// Map<K2, V2> then value's value type is V1 and key context is K1
+/// @author sgrekhov22@gmail.com
+
+// SharedOptions=--enable-experiment=patterns
+
+import "../../Utils/expect.dart";
+import "../../Utils/static_type_helper.dart";
+
+String test1() {
+  switch (<String, int>{"key1": 1}) {
+    case <Object, num>{"key1": var a}:
```

Comment:
(Damned github, I just wrote a rather detailed response, and it disappeared. Perhaps it isn't gone, and then you may see two very similar answers. ;-)

I think you'd need a `class D<X> extends C<X> { const D(); }` such that you can use a key in the pattern of the form `const D()`, which should then be inferred as `const D<B>()`.

The type arguments of the map pattern should still be `<C<B>, num>` such that `D<B>` is obtained from the context type `C<B>` in a way which involves a tiny bit of non-trivial type inference.

The initalizing expression should probably have static type `Object`, (`switch (<...>{...} as Object)`), such that we have the best possible reason to conclude that the key in the pattern was subject to type inference which is based on the actual type arguments of the map pattern.

And then the initializing expression should have a key `const D<B>()`, such that we will get a match at run time if and only if the inferred type argument of the pattern key is actually `B` as it should be.

Reply:
And it _should_ match!!! (The current version doesn't match because the initializing expression has an actual first type argument `C<A>`, which is too general when the pattern says `C<B>`, so we never reach the part where we're trying to match on a key.)

Reply:
Thank you. I've changed the test

Reply:
Thx!